Design pattern creational prototype (#418)
Codecademy/docs: content/general/concepts/what-is-the-prototype-pattern/what-is-the-prototype-pattern.md

```java
public class TvSeries implements Cloneable {
  private final String name;
  private final int seriesNo;
  // Notice one or more properties are a complex type and mutable
  private final List<Episode> episodes;

  public TvSeries(String name, int seriesNo, List<Episode> episodes) {
    this.name = name;
    this.seriesNo = seriesNo;
    this.episodes = episodes;
  }

  // Shallow copy method
  @Override
  public TvSeries clone() throws CloneNotSupportedException {
    // Cast to TvSeries as super returns Object
    return (TvSeries) super.clone();
  }

  // Deep copy
  public TvSeries deepCopy() {
    // New Episode objects are created during the copy
    return new TvSeries(this.name, this.seriesNo, this.episodes.stream()
                                                               .map(episode -> new Episode(episode.getName(), episode.getEpNo()))
                                                               .collect(Collectors.toList()));

>   }

Reviewer: The code here looks a bit awkward

Reviewer: In addition, it's important to keep this user-friendly! I feel that these code snippets are quite lengthy and difficult to follow. Is it possible to shorten these examples and still get the original point across? Every word counts!

Reviewer: In addition, it's important to keep this user-friendly! I feel that these code snippets are quite lengthy and difficult to follow. Is it possible to shorten these examples and still get the original point across? Every word counts!

Thanks for your contributions 🥇

Author: Hi,

Yes this is starting to become a little bit of an issue for me (more so in the factory.) When I first wrote this code e.g. it returned the user's last 10 watched shows regardless of whether they were films or tvshows. I then reduced it as much as I could. The issue is that these design patterns have quite a lot of moving parts to explain for a short entry definition, and I wanted to try and find an e.g. that might make some sense in the real world rather than explaining in the abstract (hence the mock repository.) Understanding how to de couple responsibilities at the class level is really important, and design patterns seem to give us a common language / way to do this. Unfortunately this leads to multiple classes. (<- I'm not totally sure this (real world thing) is the direction I should be going in. Explaining my thoughts and trying to impart conceptual ideas is still very new to me.)

With regards to the code looking a bit awkward, I would almost always prefer functional over classical (just because it can be done in one line) but sure. This copy can be done using a for each as well. No worries Ill change that after work today.

I do want this to help others and you comments are incredibly valuable to me. As my code e.g. makes sense to me by the time I write the entry, I'm never sure how clear my explanation to others are. I do want these entries to be as useful as possible.

Thank you! I'll see what I can do :)

Reviewer: Thanks for putting in the effort for a workaround!

Creating real-world scenarios is really helpful and I don't think that aspect needs to be changed. I get what you mean though about there being so many moving parts and the need to incorporate as much code content as possible. One way you can go about seeing what to keep or remove is by deciding what you can fully unpack from a code snippet and work around those fine details. Every line of code should be easily dissectable through the explanation and actual snippet.

💪

Author: Hi,

Please see my code change and Ive added more explanation. To be honest I feel I am over explaining, but happy to be guided on this.

I have also left in the comments (not sure if this is useful or not.)

I wont mark as resolved yet, as want to hear your thoughts.

Thanks again

Reviewer: I think the explanations are helpful! Really appreciate your responsiveness through this process. I'm just going to make some structural edits, double-check all the code, and then approve of this PR.

Thanks @Bernard-McGeever !

    // S2ep1 name has changed
    System.out.println(seriesTwo.getEpisodes().get(0).getName()); // Unmask
    // But, so has S1ep1 :(
    System.out.println(seriesOne.getEpisodes().get(0).getName()); // Unmask
  }
}
```

The shallow copy only copies the outermost object and references the same complex object properties in memory. As there is now one `Episode` object with two references to it, changing a value may result in unwanted behavior. Below is the same example but using the `.deepCopy()` in place of `.clone()`.

```java
public class Main {
  public static void main(String[] args) throws CloneNotSupportedException {
    // A 'new' TvSeries object is created as seriesOne
    Episode episodeOne = new Episode("Hello Friend", 13);
    final Series seriesOne = new Series("Mr Robot", 1, Collections.singletonList(episodeOne));

    // A second TvSeries object is created using a deep copy of seriesOne
    final Series seriesTwo = seriesOne.deepCopy();
    // Change the name of S2ep1
    seriesTwo.getEpisodes().get(0).setName("Unmask");
    // S2ep1 name has changed
    System.out.println(seriesTwo.getEpisodes().get(0).getName()); // Unmask
    // And s1ep1 name remains correct :)

>     System.out.println(seriesOne.getEpisodes().get(0).getName()); // Hello Friend

Reviewer: Break up code from explanation

---
Title: 'What is the Prototype Pattern?'
Description: 'The _prototype pattern_ adds cloning functionality to an objets class. After the instantiation an initial object, the prototype 
removes the need for the new keyword in creating subsequent objects at run time.'
Subjects:
- 'Computer Science'
- 'Interview Prep'

>   Tags:

Reviewer 2: ```suggestion
Tags:
```

- 'Classes'
- 'Conceptual'
- 'Constructors'
- 'Objects'

>   CatalogContent:

Reviewer 2: ```suggestion
CatalogContent:
```

- 'learn-java'
- 'paths/computer-science'
---

> The _prototype pattern_ adds cloning functionality to an objects class. After the instantiation an initial object, the prototype removes the need for the `new` keyword in creating subsequent objects at run time. New objects are created using the property values of a pre-existing object.

Reviewer 2: ```suggestion
The _prototype pattern_ adds cloning functionality to an object's class. After the instantiation an initial object, the prototype removes the need for the `new` keyword in creating subsequent objects at run time. New objects are created using the property values of a pre-existing object.
```

  private final String name;
  private final int seriesNo;
  // Notice one or more properties are a complex type and mutable
  private final List<Episode> episodes;

  public TvSeries(String name, int seriesNo, List<Episode> episodes) {
    this.name = name;
    this.seriesNo = seriesNo;
    this.episodes = episodes;
  }
    
  // Shallow copy method
  @Override
  public TvSeries clone() throws CloneNotSupportedException {
    // Cast to TvSeries as super returns Object
    return (TvSeries) super.clone();
  }

  // Deep copy
  public TvSeries deepCopy() {
    // New Episode objects are created during the copy
    return new TvSeries(this.name, this.seriesNo, this.episodes.stream()
                                                               .map(episode -> new Episode(episode.getName(), episode.getEpNo()))
                                                               .collect(Collectors.toList()));
  }
    
  // Getters and toString
}

```

```java
public class Episode {
  private String name;
  private int epNo;

  public Episode(String name, int epNo)
  {
    this.name = name;
    this.epNo = epNo;
  }

  // Getters, setters and toString
}
```

> **Note:** The properties of the episode class are mutable. This is to illustrate a potential drawback in the shallow copy method implemented in the TvSeries class.

Reviewer 2: ```suggestion
*Note:* The properties of the `Episode` class are mutable. This is to illustrate a potential drawback in the shallow copy method implemented in the `TvSeries` class.
```

> **Note:** The properties of the episode class are mutable. This is to illustrate a potential drawback in the shallow copy method implemented in the TvSeries class.

Reviewer 2: This should explicitly mention the following code block illustrates the drawbacks.

  }
}
```

The shallow copy only copies the outermost object, and references the same complex object properties in memory. As there is now one `Epesode` object with two references to it, changing a value may result in unwanted behaviour. Below is the same example but using the `deepCopy()` in place of `clone()`.

```java
public class Main {
  public static void main(String[] args) throws CloneNotSupportedException {
    // A 'new' TvSeries object is created as seriesOne
    Episode episodeOne = new Episode("Hello Friend", 13);
    final Series seriesOne = new Series("Mr Robot", 1, Collections.singletonList(episodeOne));

    // A second TvSeries object is created using a deep copy of seriesOne
    final Series seriesTwo = seriesOne.deepCopy();
    // Change the name of S2ep1
    seriesTwo.getEpisodes().get(0).setName("Unmask");
    // S2ep1 name has changed
    System.out.println(seriesTwo.getEpisodes().get(0).getName()); // Unmask
    // And s1ep1 name remains correct :)
    System.out.println(seriesOne.getEpisodes().get(0).getName()); // Hello Friend
  }
}
```

## UML Design

![UML diagram of a prototype](/media/prototype-uml.png)

## Example: Prototype

> To illustrate the _prototype pattern_, below provides a real world example, in java, depicting potential considerations for a video streaming service. For some reason, returning `Film` or `TvSeries` objects is considered to be expensive and when providing the user with their watch history, we can use the `UserRepository` to set the needed properties from cloned objects provided by a registry. 

Reviewer 2: ```suggestion
To illustrate the _prototype pattern_, below provides a real world example, in Java, depicting potential considerations for a video streaming service. For some reason, returning `Film` or `TvSeries` objects is considered to be expensive and when providing the user with their watch history, we can use the `UserRepository` to set the needed properties from cloned objects provided by a registry. 
```

> A `UserRepository` has been created with two methods; `getUserFilmHistory()` and `getUserTvSeriesHistory()`. Both simply return a list of `Strings` representing a users watch history.

Reviewer 2: ```suggestion
A `UserRepository` has been created with two methods; `getUserFilmHistory()` and `getUserTvSeriesHistory()`. Both simply return a list of `Strings` representing a user's watch history.
```

                         "Colin in Black and White", "Squid Game", "You");
  }
}
```

To model the object types to be returned, `Film` and `TvSeries` have been outlined. They both extend an abstract parent class of `Show`. This allows us to use generics in `clone()`, making the returned object automatically cast to the correct type. 

```java
public abstract class Show<T> implements Cloneable
{
  // Implements clone in abstract class to make use of generics
  @Override
  public T clone() throws CloneNotSupportedException {
    return (T) super.clone();
  }
}
```

```java
// Film extends Show<Film>, so its clone method will return a Film instance
public class Film extends Show<Film>
{
  private String title;
  private String format;
  private String thumbnailLocation;

  // Getters, setters and toString
}
```

```java
// TvSeries extends Show<TvSeries>, so its clone method will return TvSeries instance
public class TvSeries extends Show<TvSeries>
{
  private String title;
  private String format;
  private String thumbnailLocation;

  // Getters, setters and ToString
}
```

> Finally `ShowCache` has been made to add an object registry to the program. When `loadCache()` is called two template objects are created using the `new` keyword and placed in a static map. When `getShow()` is called by a client, the correct clone (taken from `showMap`) is returned, ready for its properties to be set. If the requested `showId` is not recognised, an exception is thrown.

Reviewer 2: ```suggestion
Finally `ShowCache` has been made to add an object registry to the program. When `loadCache()` is called, two template objects are created using the `new` keyword and placed in a static map. When `getShow()` is called by a client, the correct clone (taken from `showMap`) is returned, ready for its properties to be set. If the requested `showId` is not recognized, an exception is thrown.
```

  public static Hashtable<String, Show<?>> showMap = new Hashtable<>();

  // Instanciats the initial objects from which we clone from
  public static void loadCache()
  {
    final TvSeries tvSeries = new TvSeries();
    final Film film = new Film();
    showMap.put("Tv Series", tvSeries);
    showMap.put("Film", film);
  }

  // Returns clone of object stored in showMap to client
  public static Show<?> getShow(String showId) throws CloneNotSupportedException, ShowIdNotRecognisedException
  {
    // Switch statement to find out which clone is needed
    switch (showId)
    {
      case "Film":
        Film cashedFilm = (Film) showMap.get(showId);
        return cashedFilm.clone(); 
      case "Tv Series":
        TvSeries cashedTvSeries = (TvSeries) showMap.get(showId);
        return cashedTvSeries.clone();
      default:
         throw new ShowIdNotRecognisedException("Unable to get show: " + showId); 
    }
  }
}
```

```java
public class ShowIdNotRecognisedException extends Throwable
{
  private static final long serialVersionUID = 1;

  public ShowIdNotRecognisedException(final String message)
  {
    super(message);
  }
}

```

> The `Main` class both starts the program and acts as the client in this example. It asks the registry to instantiate the initial objects to clone from, before using the `UserRepository`'s data and the `getShow` method to return a list of `Film` and `TvSeries` objects representing a users previously watched shows.

Reviewer 2: ```suggestion
The `Main` class below both starts the program and acts as the client in this example. It asks the registry to instantiate the initial objects to clone from, before using the `UserRepository`'s data and the `getShow` method to return a list of `Film` and `TvSeries` objects representing a users previously watched shows.
```

> **Note:** After the instantiation of initial objects in the registry's `loadCache()`, all other `Film` and `TvSeries` objects are created through cloning functionality. The `new` keyword is absent.

Reviewer 2: ```suggestion
*Note:* After the instantiation of initial objects in the registry's `loadCache()`, all other `Film` and `TvSeries` objects are created through cloning functionality. The `new` keyword is absent.
```

---
Title: 'What is the Prototype Pattern?'

> Description: 'The _prototype pattern_ adds cloning functionality to an objets class. After the instantiation an initial object, the prototype 
> removes the need for the new keyword in creating subsequent objects at run time.'

Reviewer: ```suggestion
Description: 'The prototype pattern adds cloning functionality to an objects class.'
```

Subjects:
- 'Computer Science'
- 'Interview Prep' 
Tags:
- 'Classes'
- 'Conceptual'
- 'Constructors'
- 'Objects'
CatalogContent:
- 'learn-java'
- 'paths/computer-science'
---

The _prototype pattern_ adds cloning functionality to an object's class. After the instantiation an initial object, the prototype removes the need for the `new` keyword in creating subsequent objects at run time. New objects are created using the property values of a pre-existing object.

There are two types of copying associated with the prototype pattern. They are:

> - Shallow Copy (copies the immediate property values)
> - Deep Copy (copies the immediate values, plus any referenced object)

Reviewer: ```suggestion
- Shallow copy: Copies the immediate property values.
- Deep copy: Copies the immediate values, plus any referenced object.
```

> ## Example: Shallow Copy Vs Deep Copy

Reviewer: ```suggestion
## Example: Shallow Copy vs Deep Copy
```

> To illustrate the difference between the shallow and deep copy, a model has been provided below, consisting of two classes `TvSeries` and `Episode`. Cloning functionality has been added to TvSeries by implementing the Cloneable interface.

Reviewer: ```suggestion
To illustrate the difference between the shallow and deep copy, a model has been provided below, consisting of two classes `TvSeries` and `Episode`. Cloning functionality has been added to `TvSeries` by implementing the `Cloneable` interface.
```

```java
public class TvSeries implements Cloneable {
  private final String name;
  private final int seriesNo;
  // Notice one or more properties are a complex type and mutable
  private final List<Episode> episodes;

  public TvSeries(String name, int seriesNo, List<Episode> episodes) {
    this.name = name;
    this.seriesNo = seriesNo;
    this.episodes = episodes;
  }

  // Shallow copy method
  @Override
  public TvSeries clone() throws CloneNotSupportedException {
    // Cast to TvSeries as super returns Object
    return (TvSeries) super.clone();
  }

  // Deep copy
  public TvSeries deepCopy() {
    // New Episode objects are created during the copy
    return new TvSeries(this.name, this.seriesNo, this.episodes.stream()
                                                               .map(episode -> new Episode(episode.getName(), episode.getEpNo()))
                                                               .collect(Collectors.toList()));
  }

  // Getters and toString
}

> 

Reviewer: ```suggestion
```

```

```java
public class Episode {
  private String name;
  private int epNo;

>   public Episode(String name, int epNo)
>   {

Reviewer: ```suggestion
  public Episode(String name, int epNo) {
```

    this.name = name;
    this.epNo = epNo;
  }

>   // Getters, setters and toString

Reviewer: ```suggestion
  // Getters, setters, and toString
```

}
```

> _Note:_ The properties of the `Episode` class are mutable. This is to illustrate a potential drawback in the shallow copy method implemented in the `TvSeries` class.

Reviewer: ```suggestion
**Note:** The properties of the `Episode` class are mutable. This is to illustrate a potential drawback in the shallow copy method implemented in the `TvSeries` class.
```

The following code demonstrates a limitation of the shallow copy:

```java
public class Main {
  public static void main(String[] args) throws CloneNotSupportedException {
    // A 'new' TvSeries object is created as seriesOne
    Episode episodeOne = new Episode("Hello Friend", 13);
    final Series seriesOne = new Series("Mr Robot", 1, Collections.singletonList(episodeOne));

    // A second TvSeries object is created by shallow copying seriesOne
    final Series seriesTwo = seriesOne.clone();
    // Change the name of S2ep1
    seriesTwo.getEpisodes().get(0).setName("Unmask");
    // S2ep1 name has changed
    System.out.println(seriesTwo.getEpisodes().get(0).getName()); // Unmask
    // But, so has S1ep1 :(

>     System.out.println(seriesOne.getEpisodes().get(0).getName()); // Unmask

Reviewer: Breaking up the code snippets and with comments in between may seem like a good idea as we're documenting, but for readers, this can be a bit difficult to follow. Instead, these comments should be placed outside of the code snippet for better understanding.

  }
}
```

> The shallow copy only copies the outermost object, and references the same complex object properties in memory. As there is now one `Epesode` object with two references to it, changing a value may result in unwanted behaviour. Below is the same example but using the `deepCopy()` in place of `clone()`.

Reviewer: ```suggestion
The shallow copy only copies the outermost object and references the same complex object properties in memory. As there is now one `Episode` object with two references to it, changing a value may result in unwanted behavior. Below is the same example but using the `.deepCopy()` in place of `.clone()`.
```

```java
public class Main {
  public static void main(String[] args) throws CloneNotSupportedException {
    // A 'new' TvSeries object is created as seriesOne
    Episode episodeOne = new Episode("Hello Friend", 13);
    final Series seriesOne = new Series("Mr Robot", 1, Collections.singletonList(episodeOne));

    // A second TvSeries object is created using a deep copy of seriesOne
    final Series seriesTwo = seriesOne.deepCopy();
    // Change the name of S2ep1
    seriesTwo.getEpisodes().get(0).setName("Unmask");
    // S2ep1 name has changed
    System.out.println(seriesTwo.getEpisodes().get(0).getName()); // Unmask
    // And s1ep1 name remains correct :)
    System.out.println(seriesOne.getEpisodes().get(0).getName()); // Hello Friend
  }
}
```

## UML Design

![UML diagram of a prototype](/media/prototype-uml.png)

## Example: Prototype

> To illustrate the _prototype pattern_, below provides a real world example, in Java, depicting potential considerations for a video streaming service. For some reason, returning `Film` or `TvSeries` objects is considered to be expensive and when providing the user with their watch history, we can use the `UserRepository` to set the needed properties from cloned objects provided by a registry.

Reviewer: ```suggestion
To illustrate the _prototype pattern_, below provides a real-world example, in Java, depicting potential considerations for a video streaming service. For some reason, returning `Film` or `TvSeries` objects is considered to be expensive and when providing the user with their watch history, we can use the `UserRepository` to set the needed properties from cloned objects provided by a registry.
```

A `UserRepository` has been created with two methods; `getUserFilmHistory()` and `getUserTvSeriesHistory()`. Both simply return a list of `Strings` representing a user's watch history.

```java
public class UserRepository {
  // Returns a list of film titles
  public List<String> getUserFilmsHistory() {
    return Arrays.asList("Army of the Dead", "King Arthur: Legend of the Sword", "Total Recall", "Friday the 13th", "The General's Daughter",
                         "American Gangster", "Replicas", "Army of Thieves", "21 Jump Street", "The Harder They Fall");
  }

  // Returns a list of Tv show titles
  public List<String> getUserTvSeriesHistory() {
    return Arrays.asList("Maya and the Three", "The Thundermans", "Riverdale",  "Locke and Key", "Catching Killers", "Cocomelon", "Maid",
                         "Colin in Black and White", "Squid Game", "You");
  }
}
```

To model the object types to be returned, `Film` and `TvSeries` have been outlined. They both extend an abstract parent class of `Show`. This allows us to use generics in `clone()`, making the returned object automatically cast to the correct type.

```java

> public abstract class Show<T> implements Cloneable
> {

Reviewer: ```suggestion
public abstract class Show<T> implements Cloneable {
```

  // Implements clone in abstract class to make use of generics
  @Override
  public T clone() throws CloneNotSupportedException {
    return (T) super.clone();
  }
}
```

```java
// Film extends Show<Film>, so its clone method will return a Film instance

> public class Film extends Show<Film>
> {

Reviewer: ```suggestion
public class Film extends Show<Film> {
```

  private String title;
  private String format;
  private String thumbnailLocation;

>   // Getters, setters and toString

Reviewer: ```suggestion
  // Getters, setters, and toString
```

}
```

```java
// TvSeries extends Show<TvSeries>, so its clone method will return TvSeries instance

> public class TvSeries extends Show<TvSeries>
> {

Reviewer: ```suggestion
public class TvSeries extends Show<TvSeries> {
```

  private String title;
  private String format;
  private String thumbnailLocation;

  // Getters, setters and ToString
}
```

Finally `ShowCache` has been made to add an object registry to the program. When `loadCache()` is called two template objects are created using the `new` keyword and placed in a static map. When `getShow()` is called by a client, the correct clone (taken from `showMap`) is returned, ready for its properties to be set. If the requested `showId` is not recognized, an exception is thrown.

```java

> public class ShowCache
> {

Reviewer: ```suggestion
public class ShowCache {
```

  // A HashTable that contains the initial model object from which we clone from
  public static Hashtable<String, Show<?>> showMap = new Hashtable<>();

  // Instanciats the initial objects from which we clone from

>   public static void loadCache()
>   {

Reviewer: ```suggestion
  public static void loadCache() {
```

    final TvSeries tvSeries = new TvSeries();
    final Film film = new Film();
    showMap.put("Tv Series", tvSeries);
    showMap.put("Film", film);
  }

  // Returns clone of object stored in showMap to client

>   public static Show<?> getShow(String showId) throws CloneNotSupportedException, ShowIdNotRecognisedException
>   {

Reviewer: ```suggestion
  public static Show<?> getShow(String showId) throws CloneNotSupportedException, ShowIdNotRecognisedException {
```

    // Switch statement to find out which clone is needed
    switch (showId)
    {
      case "Film":
        Film cashedFilm = (Film) showMap.get(showId);
        return cashedFilm.clone();
      case "Tv Series":
        TvSeries cashedTvSeries = (TvSeries) showMap.get(showId);
        return cashedTvSeries.clone();
      default:
         throw new ShowIdNotRecognisedException("Unable to get show: " + showId);
    }
  }
}
```

```java

> public class ShowIdNotRecognisedException extends Throwable
> {

Reviewer: ```suggestion
public class ShowIdNotRecognisedException extends Throwable {
```

  private static final long serialVersionUID = 1;

>   public ShowIdNotRecognisedException(final String message)
>   {

Reviewer: ```suggestion
  public ShowIdNotRecognisedException(final String message) {
```

    super(message);
  }
}

```

> The `Main` class below both starts the program and acts as the client in this example. It asks the registry to instantiate the initial objects to clone from, before using the `UserRepository`'s data and the `getShow` method to return a list of `Film` and `TvSeries` objects representing a users previously watched shows.

Reviewer: ```suggestion
The `Main` class below both starts the program and acts as the client in this example. It asks the registry to instantiate the initial objects to clone from, before using the `UserRepository`'s data and the `getShow` method to return a list of `Film` and `TvSeries` objects representing a user's previously watched shows.
```

> _Note:_ After the instantiation of initial objects in the registry's `loadCache()`, all other `Film` and `TvSeries` objects are created through cloning functionality. The `new` keyword is absent.

Reviewer: ```suggestion
**Note:** After the instantiation of initial objects in the registry's `loadCache()`, all other `Film` and `TvSeries` objects are created through cloning functionality. The `new` keyword is absent.
```
